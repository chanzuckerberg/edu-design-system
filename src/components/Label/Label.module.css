@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # LABEL
\*------------------------------------*/

/**
 * Label
 * 
 * Labels can sometimes be marked up as legends for field groups (i.e. radio field).
 */
.label {
  @mixin eds-theme-typography-form-label;
  display: inline-block;
}

/**
 * Label flag
 *
 * A flag to mark whether or not a field is required or optional. Currently a flag
 * is only present for optional fields.
 */
.label__flag {
  @mixin eds-theme-typography-body-text-sm;
}

/**
 * Label after
 *
 * An empty slot for additional element to appear to the right of the label text.
 * Typically used to present tooltips by the labels.
 */
.label__after {
  display: inline-block;
  position: relative;
  top: 3px;
  margin-left: var(--eds-size-1);
}

/**
 * Visually hides the label element but is kept for screen reading accessibility.
 */
.u-is-vishidden {
  @mixin visuallyHide;
}
