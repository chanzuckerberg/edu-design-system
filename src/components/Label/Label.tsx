import clsx from 'clsx';
import type { ReactNode } from 'react';
import React from 'react';

import Text from '../Text';
import styles from './Label.module.css';

export type LabelProps = {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Indicates disabled state of the input.
   */
  disabled?: boolean;
  /**
   * HTML `for` attribute, which maps the label to an associated input `id`
   */
  htmlFor?: string;
  /**
   * HTML id for the component
   */
  id?: string;
  /**
   * Slot for node to appear directly after field label. Typically used to include a Toolip
   */
  labelAfter?: ReactNode;
  /**
   * Indicates that field is required for form to be successfully submitted.
   */
  required?: boolean;
  /**
   * The label text string
   */
  text: ReactNode;
};

/**
 * `import {Label} from "@chanzuckerberg/eds";`
 *
 * Label component used as legends for field groups (e.g., radio/checkbox fields).
 */
export const Label = ({
  className,
  disabled,
  htmlFor,
  id,
  labelAfter,
  required = true,
  text,
  ...other
}: LabelProps) => {
  const componentClassName = clsx(
    styles['label'],
    disabled && styles['label--disabled'],
    className,
  );

  return (
    <label className={componentClassName} htmlFor={htmlFor} id={id} {...other}>
      <Text as="span" preset="body-md">
        {text}
      </Text>{' '}
      {labelAfter && (
        <Text as="span" preset="body-sm">
          {labelAfter}
        </Text>
      )}
    </label>
  );
};

Label.displayName = 'Label';
