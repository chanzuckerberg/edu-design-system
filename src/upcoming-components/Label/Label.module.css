@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #LABEL
\*------------------------------------*/

/**
 * Label
 * 1) Labels can sometimes be marked up as legends for field groups (i.e. radio field)
 */
.label {
  @mixin labelStyles;
  margin-bottom: var(--eds-size-1);
}

.label--inverted {
  color: var(--eds-theme-color-text-neutral-default-inverse);
}

/**
 * Label flag
 * 1) A flag to mark whether or not a field is required
 *    or optional. Currently a flag is only present
 *    for optional fields
 */
.label__flag {
  @mixin eds-theme-typography-body-text-sm;
}

/**
 * Label after
 * 1) An empty slot for additional element to appear
 *    to the right of the label text. Typically used
 *    to present tooltips by the labels
 * 2) Media query is for IE 11 browser only for label after alignment
 */
.label__after {
  display: inline-block;
  position: relative;
  top: 3px;
  margin-left: var(--eds-size-1);
}

.u-is-vishidden {
  @mixin visuallyHide;
}
