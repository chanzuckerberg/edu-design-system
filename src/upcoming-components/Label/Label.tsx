import clsx from 'clsx';
import React, { ReactNode } from 'react';
import styles from './Label.module.css';

export interface Props {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * Toggles the visibility of the label. If hidden, the label text will still be accessible to assistive technologies
   */
  hideLabel?: boolean;
  /**
   * HTML `for` attribute, which maps the label to an associated input `id`
   */
  htmlFor?: string;
  /**
   * HTML id for the component
   */
  id?: string;
  /**
   * Inverted variation for dark backgrounds
   */
  inverted?: boolean;
  /**
   * Slot for node to appear directly after field label. Typically used to include a Toolip
   */
  labelAfter?: ReactNode;
  /**
   * String for the optional label. By default it is '(optional)'
   */
  optionalLabel?: string;
  /**
   * Indicates that field is required for form to be successfully submitted. Non-required fields will display a text "(optional)" beside the label text
   */
  required?: boolean;
  /**
   * String for the required label to add additional information if needed.
   */
  requiredLabel?: string;
  /**
   * The label text string
   */
  text?: string | boolean;
}

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * ```ts
 * import {Label} from "@chanzuckerberg/eds";
 * ```
 *
 * TODO: update this comment with a description of the component.
 */
export const Label = ({
  className,
  id,
  htmlFor,
  inverted,
  labelAfter,
  optionalLabel = '(optional)',
  text = 'Label',
  required = true,
  requiredLabel,
  hideLabel,
  ...other
}: Props) => {
  const componentClassName = clsx(
    styles['label'],
    className,
    inverted && styles['label--inverted'],
    hideLabel && styles['u-is-vishidden'],
  );

  return (
    <label className={componentClassName} htmlFor={htmlFor} id={id} {...other}>
      {text}{' '}
      {!required && (
        <span className={styles['label__flag']}>{optionalLabel}</span>
      )}
      {requiredLabel && (
        <span className={styles['label__flag']}>{requiredLabel}</span>
      )}
      {labelAfter && (
        <span className={styles['label__after']}>{labelAfter}</span>
      )}
    </label>
  );
};
