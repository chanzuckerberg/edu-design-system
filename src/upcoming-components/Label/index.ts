export { Label as default } from './Label';
