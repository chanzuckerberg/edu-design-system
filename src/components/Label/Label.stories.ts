import type { StoryObj, Meta } from '@storybook/react';
import type React from 'react';

import { Label } from './Label';

export default {
  title: 'Components/Label',
  component: Label,
} as Meta<Args>;

type Args = React.ComponentProps<typeof Label>;

export const Default: StoryObj<Args> = {
  args: {
    text: 'Label',
    required: true,
  },
};

export const Optional: StoryObj<Args> = {
  args: {
    text: 'Label',
    required: false,
  },
};
