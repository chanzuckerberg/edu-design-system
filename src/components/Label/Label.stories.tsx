import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Label, Props } from './Label';

export default {
  title: 'Atoms/Forms/Label',
  component: Label,
} as Meta;

const Template: Story<Props> = (args) => <Label {...args} />;

const InvertedTemplate: Story<Props> = (args) => (
  <div style={{ padding: '1rem', background: '#000' }}>
    <Label {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = { text: 'Label', required: true };

export const Optional = Template.bind({});
Optional.args = { text: 'Label', required: false };

export const Inverted = InvertedTemplate.bind({});
Inverted.args = { inverted: true, text: 'Label', required: true };
