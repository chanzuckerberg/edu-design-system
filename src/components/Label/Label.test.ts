import { generateSnapshots } from '@chanzuckerberg/story-utils';
import type { StoryFile } from '@storybook/testing-react';
import * as stories from './Label.stories';

describe('<Label />', () => {
  generateSnapshots(stories as StoryFile);
});
