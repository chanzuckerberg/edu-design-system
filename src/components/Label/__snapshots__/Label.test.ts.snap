// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> Default story renders snapshot 1`] = `
<label
  class="label"
>
  Label
   
</label>
`;

exports[`<Label /> LabelAfter story renders snapshot 1`] = `
<label
  class="label"
>
  Label
   
  <span
    class="text text--body-sm"
  >
    <span>
      *
    </span>
  </span>
</label>
`;
